File: spatial/R/KMLgenerate.R
#! /usr/bin/Rscript

## packages
library(data.table)
library(raster)
library(DBI)
library(dplyr) # replace RPostgreSQL
library(dbplyr) 

# Hardcoded 
test.root <- "N"
fname <- "KMLgenerate"  # KMLgenerate
kml_type <- "N"  # Type of KML (N for non-QAQC)
diam <- 0.005 / 2 # half of the pixel resolution
gcs <- "+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0"
alt.root <- NULL 
host <- NULL  # host <- "crowdmapper.org"
db.tester.name <- NULL  # db.tester.name <- "lestes"

# variables
# data(pgupw)
suppressMessages(library(rmapaccuracy)) # have to load this to get connection
coninfo <- mapper_connect(user = pgupw$user, password = pgupw$password,
                          db.tester.name = db.tester.name, 
                          alt.root = alt.root, host = host)
## File paths
log_file_path <- paste0(coninfo$dinfo["project.root"], "/log/")

## Record daemon start time
pstart_string <- paste0("KMLgenerate: Daemon starting up at ", 
                        format(Sys.time(), "%a %b %e %H:%M:%S %Z %Y"))

## Initialize csv to log database error message
log_hdr <- rbind("Error messages from KMLgenerate.R", 
                 "Time errcode errmessage",
                 "#####################################################")

### Possible conflict
dberrfname <- paste0(log_file_path, "KMLgenerate_dbase_error.log") 
if(!file.exists(dberrfname)) {
  write.table(log_hdr, file = dberrfname, sep = "", col.names = FALSE, 
              row.names = FALSE, quote = FALSE)
}
options(digit.secs = 4)  # Display milliseconds for time stamps 

# Initialize Rlog file to record daemon start time, which kml ids written & when
rlog_hdr <- rbind("Log of KMLgenerate.R start, KML ids written & times", 
                  "####################################################")
logfname <- paste0(log_file_path, fname, ".log")  # Log file name
if(!file.exists(logfname))  write(rlog_hdr, file = logfname)
# Write out daemon start stamp
write(pstart_string, file = logfname, append = TRUE)

repeat {
  # Query polling interval
  kml_polling_interval <- as.numeric(
    (tbl(coninfo$con, "configuration") %>% 
       filter(key == 'KMLPollingInterval') %>% collect())$value)
        
        # Target batch size: should be at least 500
  kml_batch_size <- as.numeric(
    (tbl(coninfo$con, "configuration") %>% filter(key == 'NKMLBatchSize') %>%
       collect())$value)
        
  # how many unmapped kmls should there be, at a minimum
  min_avail_kml <- as.numeric(
    (tbl(coninfo$con, "configuration") %>% 
       filter(key == 'MinAvailNKMLTarget') %>% collect())$value)        
        
  # how many unmapped kmls are there? 
  avail_kml_count <- as.numeric(
    tbl(coninfo$con, "kml_data") %>% 
      anti_join(
        (tbl(coninfo$con, "hit_data") %>% 
           left_join(tbl(coninfo$con, "kml_data"), by = "name") %>%
           filter(is.na(delete_time))), by = "name") %>%
      filter((kml_type == 'N') & (mapped_count == 0)) %>%
      summarise(n = n()) %>%collect())
  
  first_avail_line <- as.numeric(
    (tbl(coninfo$con, "system_data") %>% filter(key == 'firstAvailLine') %>%
       collect())$value)
        
  # Select new grid cells for conversion to kmls if N unmapped < min_avail_kml
  start_time <- Sys.time()
  if(avail_kml_count < min_avail_kml) {  
                
    # Step 1. Read the database table in order
    # Already finish the weighted sampling in create_master_grid
    xy_tabs <- coninfo$con %>% tbl("master_grid") %>%
      filter((gid >= first_avail_line) & 
               (gid <= first_avail_line + kml_batch_size - 1)) %>%
      select(id, name, avail) %>%
      head(kml_batch_size) %>% collect()
    xy_tabs <- data.table(xy_tabs)  # convert to data.table (not needed???)

    # Step 2. Update database tables
    # Update kml_data to show new names added and their kml_type
    # Quit saving the 'fwts'
    kml_new <- data.frame(name = xy_tabs$name, 
                          kml_type = rep(kml_type, nrow(xy_tabs)))
    db_insert_into(coninfo$con, table = "kml_data", values = kml_new)
                
    # Update master to show grid is no longer available for selecting/writing
    sqlrt2 <- paste0(" (", paste0("'", xy_tabs$name, "'", collapse = ","), ")")
    sql <- paste("UPDATE master_grid SET avail='F' where name in", sqlrt2)        
    dbExecute(coninfo$con, sql)
                
    # Update master_grid_counter
    count_tab <- tbl(coninfo$con, "master_grid_counter") %>% arrange(block) %>% 
      collect()
    count_tab <- data.table(count_tab, key = "counter")
    if(count_tab[, sum(counter)] == 0) {
      active_block <- count_tab[1, ]
    } else {
      active_block <- count_tab[counter < max(nrows), .SD[1]]
    }
    newcount <- active_block[, counter] + kml_batch_size
    sql <- paste0("UPDATE master_grid_counter SET counter=", newcount, 
                  " WHERE block=", active_block[, block])
    dbExecute(coninfo$con, sql)
                
    # Update the first_avail_line in configuration
    newline <- first_avail_line + kml_batch_size
    sql <- paste0("UPDATE system_data SET value=", newline,
                  " WHERE key='firstAvailLine'")
    dbExecute(coninfo$con, sql)
    
    # Step 3. Database (crude) error handling
    exception <- dbGetException(coninfo$con)  # update exceptions
    if(exception$errorNum != 0) {
      print("Error updating master_grid")  
      errors <- paste(gsub("EDT", "", Sys.time()), "  ", 
                      paste(exception, collapse = "    "))
      write(errors, file = dberrfname, append = TRUE)
      quit(status = exception$errorNum, save = "no")
    }
  }
  end_time <- Sys.time()
        
  # Write out kmlID log
  log_timestamp <- c(format(start_time, "%a %b %d %X %Y %Z"), 
                     format(end_time, "%a %b %d %X %Y %Z"))
  if(avail_kml_count < min_avail_kml) {
    write(log_timestamp[1], file = logfname, append = TRUE)
    write(as.character(kml_new$name), file = logfname, 
          append = TRUE, ncolumns = 8)
    write(log_timestamp[2], file = logfname, append = TRUE)
    write("", file = logfname, append = TRUE)
  } 
  Sys.sleep(kml_polling_interval)
}

File: spatial/python/KMLgenerate.py
# load modules
import time
import pickle
import psycopg2
import getpass
import sys
import os
from datetime import datetime as DT
import pandas as pd
import numpy as np

# save the user and password for database
'''
pgupw = {"user": "***REMOVED***", "password": "***REMOVED***"}
f = open("pgupw.data", 'wb')
pickle.dump(pgupw, f)
f.close()
'''


# Essential functions
# Function of getting database name
def getDBName(db_sandbox_name="AfricaSandbox", db_product_name="Africa",
              db_tester_name=None, alt_root=None):
    euser = getpass.getuser()
    uname = None
    sandbox = False
    if euser == "sandbox":
        sandbox = True
        uname = "sandbox"
    elif euser == "mapper":
        sandbox = False
        uname = "mapper"
    elif db_tester_name is not None:
        if euser == db_tester_name:
            sandbox = True
            uname = "sandbox"
    else:
        sys.exit("Any script must run under sandbox or mapper user!")

    # project root path
    if alt_root is not None:
        project_root = alt_root
    else:
        project_root = "/home/" + uname + "/afmap"

    # DB Names
    if sandbox:
        db_name = db_sandbox_name
    else:
        db_name = db_product_name

    return {"db_name": db_name, "project_root": project_root}


# Function of connecting database
def mapper_connect(user, password, host=None, db_tester_name=None, alt_root=None):
    # Pull working environment
    dinfo = getDBName(db_tester_name=db_tester_name, alt_root=alt_root)

    # Paths and connections
    con = psycopg2.connect(host=host, database=dinfo["db_name"], user=user, password=password)

    return {"con": con, "dinfo": dinfo}


# Function of generating regular KMLs
def RegKML_gen(fname=None, kml_type=None, alt_root=None,
               host=None, db_tester_name=None, pgupw=None):
    fname = fname
    kml_type = kml_type
    alt_root = alt_root
    host = host
    db_tester_name = db_tester_name
    pgupw = pgupw

    # Connect to the database
    coninfo = mapper_connect(user=pgupw["user"], password=pgupw["password"],
                             db_tester_name=db_tester_name,
                             alt_root=alt_root, host=host)

    # File paths
    log_file_path = coninfo["dinfo"]["project_root"] + "/log/"

    # Record daemon start time
    pstart_string = pd.DataFrame(["KMLgenerate: Daemon starting up at " + str(DT.now())])

    # Initialize csv to log database error message
    log_hdr = pd.DataFrame(["Error messages from KMLgenerate.py", "Time errcode errmessage",
                            "#####################################################"])

    # Possible conflict
    dberrfname = log_file_path + "KMLgenerate_dbase_error.log"
    if not os.path.isfile(dberrfname):
        log_hdr.to_csv(dberrfname, index=False, index_label=False, header=False)

    np.set_printoptions(precision=4)  # Display milliseconds for time stamps

    # Initialize Rlog file to record daemon start time, which kml ids written & when
    rlog_hdr = pd.DataFrame(["Log of KMLgenerate.py start, KML ids written & times",
                             "#####################################################"])
    logfname = log_file_path + fname + ".log"  # log file name
    if not os.path.isfile(logfname):
        rlog_hdr.to_csv(logfname, index=False, index_label=False, header=False)

    # Write out daemon start stamp
    pstart_string.to_csv(logfname, mode="a", index=False, index_label=False, header=False)

    while True:
        curs = coninfo["con"].cursor()

        # Query polling interval
        curs.execute("select value from configuration where key = 'KMLPollingInterval';")
        kml_polling_interval = int(curs.fetchone()[0])

        # Target batch size: should be at least 500
        curs.execute("select value from configuration where key = 'NKMLBatchSize';")
        kml_batch_size = int(curs.fetchone()[0])

        # how many unmapped kmls should there be, at a minimum
        curs.execute("select value from configuration where key = 'MinAvailNKMLTarget';")
        min_avail_kml = int(curs.fetchone()[0])

        # how many unmapped kmls are there?
        curs.execute(
            "select count(*) from kml_data k where not exists (select true from hit_data h where h.name = k.name "
            "and delete_time is null) and kml_type = 'N' and mapped_count = 0")
        avail_kml_count = int(curs.fetchone()[0])

        # Get the anchor of loading data
        curs.execute("select value from system_data where key = 'firstAvailLine';")
        first_avail_line = int(curs.fetchone()[0])

        # Select new grid cells for conversion to kmls if N unmapped < min_avail_kml
        if avail_kml_count < min_avail_kml:

            start_time = str(DT.now())

            # Step 1. Poll the database to see which grid IDs are still available
            curs.execute("select id, name from master_grid where avail = 'T' and gid >= " +
                         str(first_avail_line) + " and gid <= " +
                         str(first_avail_line + kml_batch_size - 1))
            xy_tabs = curs.fetchall()

            try:
                # Step 2. Update database tables
                # Update kml_data to show new names added and their kml_type
                # Update master to show grid is no longer available for selecting/writing
                for row in xy_tabs:
                    xy_tab = row + (kml_type, 0)
                    insert_query = "insert into kml_data (gid, name, kml_type, mapped_count) values (%s, %s, %s, %s);"
                    curs.execute(insert_query, xy_tab)
                    curs.execute("update master_grid set avail='F' where name = '%s'" % row[1])
                coninfo["con"].commit()

                # Update the first_avail_line in configuration
                new_line = first_avail_line + kml_batch_size
                curs.execute("update system_data set value='%s' where key='firstAvailLine'" % new_line)
                coninfo["con"].commit()

            except psycopg2.DatabaseError, err:
                print "Error updating database"
                error = str(DT.now()) + " " + str(err)
                file(dberrfname, "a").write(error + os.linesep)
                sys.exit(1)
            finally:
                if coninfo["con"]:
                    coninfo["con"].close()

            end_time = str(DT.now())

            # Write out kmlID log
            file(logfname, "a").write(start_time + os.linesep)
            names = pd.DataFrame(xy_tabs, columns=["id", "name"])["name"]
            names = zip(*[iter(names)] * 8)
            for i, row in enumerate(names):
                file(logfname, "a").write(str(row) + os.linesep)
            file(logfname, "a").write(end_time + os.linesep)

        time.sleep(kml_polling_interval)


# Function of listening to cvml and generating KMLs
def cvmlKML_gen(fname=None, kml_type=None, alt_root=None,
                host=None, db_tester_name=None, pgupw=None):
    pass


# Main function
def main(fname=None, kml_type=None, alt_root=None,
         host=None, db_tester_name=None, pgupw=None):
    if True:
        RegKML_gen(fname, kml_type, alt_root,
                   host, db_tester_name, pgupw)
    else:
        cvmlKML_gen(fname, kml_type, alt_root,
                    host, db_tester_name, pgupw)


if __name__ == "__main__":
    # Hardcoded
    file_name = "KMLgenerate"  # KMLgenerate
    ktype = "N"  # Type of KML (N for non-QAQC)
    alter_root = None
    db_host = None
    db_user_name = "lsong"
    f = open("pgupw.data", "rb")  # Read user name and password for DB
    pg_pw = pickle.load(f)
    f.close()

    main(fname=file_name, kml_type=ktype, alt_root=alter_root,
         host=db_host, db_tester_name=db_user_name, pgupw=pg_pw)
